File: tileSolver.py
BLANK = 0
UP = 'up'
DOWN = 'down'
LEFT = 'left'
RIGHT = 'right'


class Board:
    def __init__(self, layout, size = 3):
        self.size = size
        self.layout = layout
        self.solved = list(range(1,(size*size)))
        self.solved.append(0)
        
        
    def buildBoard(self, layout):
        board = []
        for i in layout:
            board.append(i)
        
        return board
        
    def printBoard(self):
        board = self.board
        for y in range(self.size): # Iterate over each row.
            for x in range(self.size): # Iterate over each column.
                    print(str(board[y * self.size + x]) + ' ', end='')
            print() # Print a newline at the end of the row.

    def findEmptyTile(self):
        return self.board.index(BLANK)

    def getValidMoves(self, prevMove = None):
        emptyTile = self.findEmptyTile()
        validMoves = []

        emptyTileX = emptyTile % self.size
        emptyTileY = (emptyTile - emptyTileX) / self.size

        if emptyTileY < self.size - 1 and prevMove != DOWN:
            validMoves.append(UP)  

        if emptyTileY > 0 and prevMove != UP:
            validMoves.append(DOWN)

        if emptyTileX < self.size - 1 and prevMove != RIGHT:
            validMoves.append(LEFT)

        if emptyTileX > 0 and prevMove != LEFT:
            validMoves.append(RIGHT)
            

        return self.heuristicFunction(validMoves)
        #return validMoves

    def heuristicFunction(self, moves):
        emptyTile = self.findEmptyTile()

        sortedMoves = []

        for move in moves:
            tile = self.findNeighbouringTile(move)

            #Finds the relative location for the given value of the tile versus its location on the board
            relativePos = self.board[tile] - tile

            #Prioritize moves that push tiles towards their goal location
            if relativePos < 0 and (move == UP or move == LEFT):
                sortedMoves.insert(0, move) 
            elif relativePos > 0 and (move == DOWN or move == RIGHT):
                sortedMoves.insert(0, move) 
            else:
                sortedMoves.append(move)

        return sortedMoves

    def findNeighbouringTile(self,move):
        emptyTile = self.findEmptyTile()

        if move == UP:
            tile = emptyTile + self.size
        elif move == DOWN:
            tile = emptyTile - self.size
        elif move == LEFT:
            tile = emptyTile + 1
        elif move == RIGHT:
            tile = emptyTile - 1

        return tile

    def makeMove(self, move):
        emptyTile = self.findEmptyTile()

        swappedTile = self.findNeighbouringTile(move)

        self.board[emptyTile], self.board[swappedTile] = self.board[swappedTile], self.board[emptyTile]

    def attemptMove(self, moveList, movesRemaining, prevMove):
        # Base Case - Ran out of moves.
        if movesRemaining < 0:
            return False
        
        # Base Case - Solved the puzzle.
        if self.board == self.solved:
            return True
            
        # Recursive Case - Attempt each of the moves from getValidMoves:
        for move in self.getValidMoves(prevMove):
            self.makeMove(move)
            moveList.append(move)

            if self.attemptMove(moveList, movesRemaining - 1, move):
                self.undoMove(move) # Undo's last move
                return True

            # Undo the move to set up for the next move:
            self.undoMove(moveList.pop())

        return False 

    def undoMove(self, move):
        if move == UP:
            self.makeMove(DOWN)
        elif move == DOWN:
            self.makeMove(UP)
        elif move == LEFT:
            self.makeMove(RIGHT)
        elif move == RIGHT:
            self.makeMove(LEFT)


    def solveBoard(self, maxMoves):
        solution = []
        solvedBoard = self.attemptMove(solution, maxMoves, None)

        if solvedBoard:
            self.printBoard()
            for move in solution:
                print()
                print("Move: ", move)
                self.makeMove(move)
                self.printBoard()
                print()

        self.solution = solution
        return solvedBoard

        
        
def readFile():
    path = input("Please enter file path: ")
    try:
        with open(path) as reader:
            values = reader.read().split()
            output = []
            for i in values:
                if i.isdigit():
                    output.append(int(i))
                else:
                    raise Exception("Invalid Input Value")
            return output
    except FileNotFoundError:
        raise Exception("File could not be opened")

def main():
    try:
        input = readFile()
        
        if input != None:
            size = int(input.pop(0))
            
            if size*size == len(input):

                dup = input.copy()  
                for i in input:
                    dup.pop(0)
                    for d in dup:
                        if i == d:
                            raise Exception("Invalid Input - Duplicate Values")

                print(input)
                board = Board(input, size)
                board.board = board.buildBoard(board.layout)
                
                    
                maxMoves = 10
                while(True):
                    print("Attempting with maximum ", maxMoves, " moves.")
                    if board.solveBoard(maxMoves):
                        print("Solved with ", len(board.solution), " moves.")
                        print(board.solution)
                        break
                    
                    maxMoves += 1
            else:
                raise Exception("Invalid Input - Number of Values")
    except Exception as exception:
        print(exception)

            
    
if __name__ == "__main__": # this is for specifc use when runing as a script
    main()

File: tileSolver.hs
import System.IO
import Control.Exception
import Text.Read
import Data.Maybe
import Prelude hiding (Word)
import Data.Function (on)
import Data.List (minimumBy, delete)
import Control.Monad (guard)


--code for reading input from file

readInValues :: IO [Int]
readInValues = do
    putStrLn "Please enter file path"
    path <- getLine
    file <- try (readFile path) :: IO (Either SomeException String)
    case file of
        Left exception -> do
            putStrLn "File could not be opened"
            return []
        Right contents -> do
            let inVal = mapMaybe readMaybe (words contents)
            let values = checkSize inVal
            if (values == [])
            then do
                putStrLn "Invalid input argumets"
                return []
            else
                return values
   
checkSize :: [Int] -> [Int]
checkSize [] =  []
checkSize (size:list) = if (size*size == length list && allUnique list)
                    then (size:list)
                    else []
                    
allUnique :: [Int] -> Bool
allUnique [] = True
allUnique (x:xs) = notElem x xs && allUnique xs
                    
--code for checking if puzzle is solveable
    
--Looks through each number in list starting from 1
--Counts numbers higher than current number that appear earlier in the list
getInversions :: [Int] -> Int -> Int -> Int
getInversions list size currNum =
    if currNum == (size * size)
    then 0
    
    else
    length (filter (>currNum) (take (findElem list currNum) list))
    + getInversions list size (currNum + 1)

    
--Empty tile is 0 when calling findRow
--Starting Number is 1 when calling getInversions
-- If Even Size - Inversions + empty tile row must be odd to be solvable
-- If Odd Size - Inversions must be even to be solvable
isSolvable :: [Int] -> Bool
isSolvable (size:list)
  | size `mod` 2 == 0 = ((getInversions list size 1) + (findRow list 0 size)) `mod` 2 /= 0
  | otherwise         = (getInversions list size 1) `mod` 2 == 0



--Finds desired number in list of numbers
findElem :: [Int] -> Int -> Int
findElem (x:xs) num
  | x == num  = 0
  | otherwise = 1 + findElem xs num
  
--Finds the number row from the number list and row size
findRow :: [Int] -> Int -> Int -> Int
findRow list num size = floor $ fromIntegral (findElem list num) / fromIntegral size

-- Code to solve the puzzle

-- Input: Size
generateSolution :: Int -> [Int]
generateSolution size = [1..(size^2 - 1)] ++ [0]

-- Input: Board -> Size -> prevMove
getValidMoves board size prevMove = let
      empty = findEmptyTile board 0
      in
      filter (not . null) [getValidMove empty size prevMove validMove | validMove <- ["UP", "DOWN", "LEFT", "RIGHT"]]

-- Input: Empty Tile -> Size -> Previous Move -> Valid Move
getValidMove :: Int -> Int -> [Char] -> [Char] -> [Char]
getValidMove empty size prevMove validMove
      | (y < (size - 1)) && prevMove /= "DOWN" && validMove == "UP" = "UP"
      | (y > 0) && prevMove /= "UP" && validMove == "DOWN" = "DOWN"
      | (x < (size - 1)) && prevMove /= "RIGHT" && validMove == "LEFT" = "LEFT"
      | (x > 0) && prevMove /= "LEFT" && validMove == "RIGHT" = "RIGHT"
      | otherwise = []
      where
      x = empty `mod` size
      y = (empty - x) `div` size

-- Input: Board -> Count -> Size
printBoard :: [Int] -> Int -> Int -> IO()
printBoard [] _ _ = putChar '\n'
printBoard (a:b) count size
      | count == size = do
      putChar '\n'
      printBoard (a:b) 0 size

      |otherwise =
      do
      putStr(show a)
      if (a < 10)
      then putStr "  "
      else putStr " "
      printBoard b (count+1) size

-- Input: Board -> Move List -> Size
printSolution _ [] _ = return ()
printSolution board (x:xs) size = do
      putStrLn x
      let newBoard = makeMove board x size
      printBoard newBoard 0 size
      printSolution newBoard xs size

-- Input: Board -> Move -> Size
makeMove :: [Int] -> [Char] -> Int -> [Int]
makeMove board move size
      | a < b = swapTiles a b board
      | otherwise = swapTiles b a board
      where
      a = findEmptyTile board 0
      b = findNeighbouringTile a move size

-- Input: Board -> Count = 0
findEmptyTile :: [Int] -> Int -> Int
findEmptyTile (a:b) count
      | a == 0 = count
      | otherwise = findEmptyTile b (count + 1)

-- Input: Empty Tile -> Move -> Size
findNeighbouringTile :: Int -> [Char] -> Int -> Int
findNeighbouringTile emptyTile move size
      | move == "UP" = emptyTile + size
      | move == "DOWN" = emptyTile - size
      | move == "LEFT" = emptyTile + 1
      | move == "RIGHT" = emptyTile - 1

-- Input: Index 1 -> Index 2 -> Board
swapTiles :: Int -> Int -> [a] -> [a]
swapTiles i j xs = let
      elemI = xs !! i
      elemJ = xs !! j
      left = take i xs
      middle = take (j - i - 1) (drop (i + 1) xs)
      right = drop (j + 1) xs
      in  left ++ [elemJ] ++ middle ++ [elemI] ++ right

-- Input: Board -> Size -> Move List -> Moves Remaining -> Previous Move
attemptMove :: [Int] -> Int -> [[Char]] -> Int -> [Char] -> [[Char]]
attemptMove board size moveList movesRemaining prevMove
      | board == (generateSolution size) = moveList
      | movesRemaining <= 0 = []
      | otherwise = do
            move <- (getValidMoves board size prevMove)
            let newBoard = makeMove board move size
            attemptMove newBoard size (moveList ++ [move]) (movesRemaining - 1) move

-- Input: (Size:Board) -> Max Number of Moves
solveBoard :: [Int] -> Int -> IO()
solveBoard (size:board) maxMoves = do
      putStrLn ("Attempting with maximum " ++ show maxMoves ++ " moves")
      if (board == (generateSolution size))
      then do
            putStrLn ("Solved in 0 moves")
            printBoard board 0 size
      else
            if (moveList /= [])
            then do
                  putStrLn ("Solved in "  ++ show maxMoves ++  " moves")
                  printBoard board 0 size
                  (printSolution board moveList size)
                  print (moveList)
            else
                  solveBoard (size:board) (maxMoves + 1)

            where moveList = (attemptMove board size [] maxMoves [])

--main driver
main :: IO()
main = do
    values <- readInValues
    if (values /= [])
    then
        if (isSolvable values)
        then
            solveBoard values 1
        else
            putStrLn "puzzle not solvable"
    else return ()
        
    

